crates/ln-dlc-node/src/networking.rs

```diff
@@ -0,0 +1,89 @@
+use crate::node::NodeInfo;
+use futures::future;
+use lightning::ln::peer_handler::APeerManager;
+use lightning::ln::peer_handler::SocketDescriptor;
+use std::future::Future;
+use std::ops::Deref;
+use tracing::debug;
+
+#[cfg(feature = "ln_net_axum_ws")]
+pub mod axum;
+#[cfg(feature = "ln_net_tcp")]
+pub mod tcp;
+#[cfg(feature = "ln_net_ws")]
+mod tungstenite;
+
+// TODO we can let the connect functions be async as well
+
+#[allow(clippy::diverging_sub_expression, unused_variables, unreachable_code)] // From the panic!() below
+pub async fn connect_outbound<PM: Deref + 'static + Send + Sync + Clone>(
+    peer_manager: PM,
+    peer: NodeInfo,
+) -> Option<impl Future<Output = ()>>
+where
+    PM::Target: APeerManager<Descriptor = DynamicSocketDescriptor>,
+{
+    if peer.is_ws {
+        debug!("Connecting over WS");
+
+        #[cfg(not(feature = "ln_net_ws"))]
+        let ws: Option<future::Either<future::Ready<()>, _>> =
+            panic!("Cannot connect outbound over WS when ln_net_ws is not enabled");
+
+        #[cfg(feature = "ln_net_ws")]
+        let ws = tungstenite::connect_outbound(peer_manager, peer)
+            .await
+            .map(future::Either::Left);
+
+        ws
+    } else {
+        debug!("Connecting over TCP");
+
+        #[cfg(not(feature = "ln_net_tcp"))]
+        let tcp: Option<future::Either<_, future::Ready<()>>> =
+            panic!("Cannot connect outbound over TCP when ln_net_tcp is not enabled");
+
+        #[cfg(feature = "ln_net_tcp")]
+        let tcp = tcp::connect_outbound(peer_manager, peer.pubkey, peer.address)
+            .await
+            .map(future::Either::Right);
+
+        tcp
+    }
+}
```

Comment:
🔧 I think you should be able to get rid of all of the `allow`s if you use the `cfg!` macro. You could handle all the valid combinations explicitly and panic for all the invalid cases in a single `else` branch.

Reply:
I don't think we can use the cfg! macro here, as that resolves to a runtime bool. Hence, we'd end up with a call to `tcp::connect_outbound` still in the final code, which would not exist if the `ln_net_tcp` feature were not enabled.

I tried with the `cfg_if` crate but it doesn't support expression returns so regardless it's kind of the same.

However, one way we may be able to solve it is by explicit `return` statements sprinkled around, if you'd prefer?

Reply:
I tried this
```rs
    if peer.is_ws {
        debug!("Connecting over WS");

        #[cfg(not(feature = "ln_net_ws"))]
        return panic!("Cannot connect outbound over WS when ln_net_ws is not enabled");

        #[cfg(feature = "ln_net_ws")]
        tungstenite::connect_outbound(peer_manager, peer)
            .await
            .map(future::Either::Left)
    } else {
        debug!("Connecting over TCP");

        #[cfg(not(feature = "ln_net_tcp"))]
        return panic!("Cannot connect outbound over TCP when ln_net_tcp is not enabled");

        #[cfg(feature = "ln_net_tcp")]
        tcp::connect_outbound(peer_manager, peer.pubkey, peer.address)
            .await
            .map(future::Either::Right)
    }
    ```
    which is pleasing but doesn't give Rust enough info to infer the type of `Either` - which we can't know unless we know the features that are enabled, which is a bit of a pain.

Reply:
Okay, thanks for giving it a go! We can leave it. If I find time after we merge this I will play around with it and report back.